packages/server/src/migrations/functions/tableSettings.ts

```diff
@@ -0,0 +1,143 @@
+import { getScreenParams } from "../../db/utils"
+import { Screen } from "@budibase/types"
+import { makePropSafe as safe } from "@budibase/string-templates"
+/**
+ * Date:
+ * November 2022
+ *
+ * Description:
+ * Update table settings to use actions instead of links. We do not remove the
+ * legacy values here as we cannot guarantee that their apps are up-t-date.
+ * It is safe to simply save both the new and old structure in the definition.
+ *
+ * Migration 1:
+ * Legacy "linkRows", "linkURL", "linkPeek" and "linkColumn" settings on tables
+ * and table blocks are migrated into a "Navigate To" action under the new
+ * "onClick" setting.
+ *
+ * Migration 2:
+ * Legacy "titleButtonURL" and "titleButtonPeek" settings on table blocks are
+ * migrated into a "Navigate To" action under the new "onClickTitleButton"
+ * setting.
+ */
+export const run = async (appDb: any) => {
+  // Get all app screens
+  let screens: Screen[]
+  try {
+    screens = (
+      await appDb.allDocs(
+        getScreenParams(null, {
+          include_docs: true,
+        })
+      )
+    ).rows.map((row: any) => row.doc)
+  } catch (e) {
+    // sometimes the metadata document doesn't exist
+    // exit early instead of failing the migration
+    console.error("Error retrieving app metadata. Skipping", e)
+    return
+  }
+
+  // Recursively update any relevant components and mutate the screen docs
+  screens.forEach((screen: any) => {
+    migrateTableSettings(screen.props, screen)
+
+    // Save screen if we updated it
+    if (screen.touched) {
+      delete screen.touched
+      appDb.put(screen)
+      console.log(
+        `Screen ${screen.routing?.route} contained table settings which were migrated`
+      )
+    }
+  })
+}
+
+// Recursively searches and mutates a screen doc to migrate table component
+// and table block settings
+const migrateTableSettings = (component: any, screen: any) => {
```

Comment:
Could we type `Screen` here. 

NAB: 
Would be awesome to have a base `Component` type. I could see this extending down the line where we have type safety on the components for doing migrations like this. 
```ts
interface Component {}
interface TableComponent extends Component {}

// with useful type assertion functions
const isTableComponent = (component: Component): component is TableComponent => {
   return component._component.endsWith("/table")
}
```

Reply:
I had it typed originally, but then because I needed to use a temporary `touched` property on the screen to only save updated screens, that of course broke the typing since it doesn't exist on the type. It doesn't use the screen for anything other than setting the `touched` property anyway so we aren't missing out on anything.

I agree about the component typing - it would be nice to have.

Reply:
I can refactor this to just use tail end recursion and return a bool stating whether or not a change needs made - then I don't need to pass in the screen at all actually, and it's probably cleaner.

Reply:
Ah makes sense. Another option that could have been used is a locally defined utility type

```ts
type MigratingScreen = Screen & { touched: boolean }
```

But your new approach sounds good 👍